base/abstractarray.jl

```diff
@@ -86,8 +128,48 @@ linearindexing(A::AbstractArray, B::AbstractArray...) = linearindexing(linearind
 linearindexing(::LinearFast, ::LinearFast) = LinearFast()
 linearindexing(::LinearIndexing, ::LinearIndexing) = LinearSlow()
 
+abstract IndicesBehavior
+immutable IndicesStartAt1  <: IndicesBehavior end   # indices 1:size(A,d)
+immutable IndicesUnitRange <: IndicesBehavior end   # arb UnitRange indices
+immutable IndicesList      <: IndicesBehavior end   # indices like (:cat, :dog, :mouse)
+
+indicesbehavior(A::AbstractArray) = indicesbehavior(typeof(A))
+indicesbehavior{T<:AbstractArray}(::Type{T}) = IndicesStartAt1()
+indicesbehavior(::Number) = IndicesStartAt1()
+
+abstract IndicesPerformance
+immutable IndicesFast1D <: IndicesPerformance end  # indices(A, d) is fast
+immutable IndicesSlow1D <: IndicesPerformance end  # indices(A) is better than indices(A,d)
+
+indicesperformance(A::AbstractArray) = indicesperformance(typeof(A))
+indicesperformance{T<:AbstractArray}(::Type{T}) = IndicesFast1D()
+
+"""
+    shape(A)
+
+Returns a tuple specifying the "shape" of array `A`. For arrays with
+conventional indexing (indices start at 1), this is equivalent to
+`size(A)`; otherwise it is equivalent to `incides(A)`.
+"""
```

Comment:
I like it, but

> this would be invisible to the end user

might be too much to hope for. Say the user is looking for particular entries in the array that satisfy some kind of criterion: the user might allocate an `Array{Int}(0)` and then `push!` onto it. That would work if we defined `convert(::Type{Int}, i::LinearIndex)`, but then one couldn't use them for indexing later. So the user would at least have to know to allocate an `Array{LinearIndex}(0)`.

There's also the question of what functions `LinearIndex` should have defined. Certainly basic arithmetic, but what about, e.g., `sqrt`?